Don't always escape allocations when they're written to the stack in StoreBarrierInsertionPhase

saambarati opened the pull request:
#### b5a538c2f23dc7ecb423898320c72f8befeb4d24
<pre>
Don&apos;t always escape allocations when they&apos;re written to the stack in StoreBarrierInsertionPhase
<a href="https://bugs.webkit.org/show_bug.cgi?id=242157">https://bugs.webkit.org/show_bug.cgi?id=242157</a>

Reviewed by Yusuke Suzuki.

This is an optimization after my change in 251925@main. It turns out
that we end up writing allocations out to the stack frequently in
Speedometer2, and we were always considering this an escape. This lead
us to have some small slowdowns in a few subtests.

This patch makes it so we do a basic escape analysis on these allocations
being written to the stack. Basically, the write of the allocation to the
stack is no longer an escape. Instead, we track it for the current epoch.
Only if we read from that stack slot in the same epoch do we escape the allocation.

* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:

Canonical link: <a href="https://commits.webkit.org/251979@main">https://commits.webkit.org/251979@main</a>
</pre>

Constellation commented on Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
```diff
@@ -431,17 +435,50 @@ class StoreBarrierInsertionPhase : public Phase {
-                if (writes.overlaps(Heap) || writes.overlaps(Stack)) {
-                    auto escape = [] (Node* node) {
+                auto readFunc = [&] (const AbstractHeap& heap) {
+                    if (heap.kind() == Stack) {
```
Don't we need to have `read(World)` support?

saambarati replied:
World will overlap with `heap.overlaps(Heap)` below. Maybe I can make the code a bit more obvious, though.

saambarati replied:
sorry, I thought this comment was about some other code. You're right, let me fix this.

Constellation commented on Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
```diff
@@ -431,17 +435,50 @@ class StoreBarrierInsertionPhase : public Phase {
+                    if (heap.kind() == Stack) {
+                        wroteHeapOrStack = true;
+                        if (!heap.payload().isTop())
+                            preciseStackWrite = heap;
```
Can it be a Vector?
For example, `LoadVarargs` / `ForwardVarargs` are doing `write(AbstractHeap(Stack, ...))` multiple times.

saambarati replied:
Yeah, but we're only considering this for `PutStack`/`SetLocal`. Let me add a RELEASE_ASSERT that for these nodes we only see a single stack write.

Constellation commented on Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
```diff
@@ -431,17 +435,50 @@ class StoreBarrierInsertionPhase : public Phase {
+                        }
+
+                        for (const auto& heap : toRemove)
+                            potentialStackEscapes.remove(heap);
```
Let's use `potentialStackEscapes.removeIf` since it is more efficient.

saambarati replied:
Sounds good!

Constellation reviewed (approved):
r=me with comments.

saambarati committed:
Don't always escape allocations when they're written to the stack in StoreBarrierInsertionPhase
https://bugs.webkit.org/show_bug.cgi?id=242157

Reviewed by Yusuke Suzuki.

This is an optimization after my change in 251925@main. It turns out
that we end up writing allocations out to the stack frequently in
Speedometer2, and we were always considering this an escape. This lead
us to have some small slowdowns in a few subtests.

This patch makes it so we do a basic escape analysis on these allocations
being written to the stack. Basically, the write of the allocation to the
stack is no longer an escape. Instead, we track it for the current epoch.
Only if we read from that stack slot in the same epoch do we escape the allocation.

* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:

Canonical link: https://commits.webkit.org/251979@main
